src/NetworkName.js

```diff
@@ -1,60 +1,175 @@
- * @property {ClientNetworkName} Mainnet
- * @property {ClientNetworkName} Testnet
- * @property {ClientNetworkName} Previewnet
- */
-
-/**
- * @type {NetworkNameType}
- */
-const NetworkName = {
-    Mainnet: "mainnet",
-    Testnet: "testnet",
-    Previewnet: "previewnet",
-};
-
-/** @type {[string, string, string]} */
-export const _networkIds = ["0", "1", "2"];
-
-/**
- * @param {string} networkName
- * @returns {string}
- */
-export function _ledgerIdToLedgerId(networkName) {
-    switch (networkName) {
-        case NetworkName.Mainnet:
-            return _networkIds[0];
-        case NetworkName.Testnet:
-            return _networkIds[1];
-        case NetworkName.Previewnet:
-            return _networkIds[2];
-        default:
-            throw new Error(
-                `unrecognized network name, expected "mainnet", "testnet", or "previewnet"`
-            );
+const DEFAULT_ERROR = "Default case reached for: ";
+
+export default class NetworkName {
+    /**
+     * @hideconstructor
+     * @internal
+     * @param {number|string} networkId
+     */
+    constructor(networkId) {
+        /**
+         * @readonly
+         * @type {number}
+         */
+        this._networkId =
+            typeof networkId == "string" ? parseInt(networkId) : networkId;
+
+        /**
+         * @readonly
+         * @type {string}
+         */
+        this._networkName = NetworkName.networkNameFromId(this._networkId);
+
+        Object.freeze(this);
+    }
+
+    /**
+     * @returns {number}
+     */
+    get networkId() {
+        return this._networkId;
+    }
+
+    /**
+     * @returns {string}
+     */
+    get networkName() {
+        return this._networkName;
+    }
+
+    /**
+     * @returns {string}
+     */
+    toString() {
+        return this.networkId.toString();
+    }
+
+    /**
+     * @param {string} networkNameOrId
+     * @returns {NetworkName}
+     */
+    static fromString(networkNameOrId) {
+        switch (networkNameOrId) {
+            case "0":
+            case NetworkName.MAINNET:
+                return new NetworkName(0);
+            case "1":
+            case NetworkName.TESTNET:
+                return new NetworkName(1);
+            case "2":
+            case NetworkName.PREVIEWNET:
+                return new NetworkName(2);
+            default:
+                throw new Error(DEFAULT_ERROR.concat("fromString"));
+        }
+    }
+
+    /**
+     * @param {string|number} networkId
+     * @returns {Uint8Array}
+     */
+    static networkIdToProtobuf(networkId) {
+        return new Uint8Array(
+            typeof networkId != "string" ? [networkId] : [parseInt(networkId)]
+        );
+    }
+
+    /**
+     * @returns {Uint8Array}
+     */
+    toProtobuf() {
+        // TODO
+        // rename?
+        return new Uint8Array([this._networkId]);
+    }
+
+    /**
+     * @param {Uint8Array} networkName
+     * @returns {NetworkName}
+     */
+    static fromProtobuf(networkName) {
+        // TODO
```

Comment:
thoughts? @danielakhterov

Reply:
I believe this was before we talked, but we don't really need `NetworkName.[to|from]Protobuf()` However, if you'd like to implement it it's fine. The type this should encode/decode to should be `Uint8Array`.